File: typescript/src/nuc/BLOCK$CLASS.ts
// src/nuc/BLOCK$CLASS.ts

import BLOCK$INSTANCE from "./BLOCK$INSTANCE";
import Instruction from "../Instruction";
import NODE from "./NODE";
import Scope from "../Scope";
import { v4 as uuid } from "uuid";

class BLOCK$CLASS extends NODE {
  statements: any[];
  class!: {
    instances: Record<string, Record<string, unknown>>;
    declarations: Record<string, BLOCK$CLASS>;
  };
  id!: string;
  type: string;

  constructor(key: unknown) {
    super(typeof key === "string" ? key : String(key));
    this.statements = [];
    this.type = "CLASS";
  }

  run(scope: Scope): NODE[] | null {
    const instances: Record<string, unknown>[] = scope.$instance
      ? [scope.$instance as Record<string, unknown>]
      : Object.values(this.class.instances);

    const instructions: NODE[] = [];

    for (const instance of instances) {
      const statement = new BLOCK$INSTANCE(uuid()) as unknown as {
        id: string;
        class?: {
          instances: Record<string, Record<string, unknown>>;
          declarations: Record<string, BLOCK$CLASS>;
        };
        instance?: Record<string, unknown>;
        statements?: any[];
        declaration?: BLOCK$CLASS;
      };

      statement.id = uuid();
      statement.class = this.class;
      statement.instance = instance;
      statement.statements = this.statements;
      statement.declaration = this;

      const parentScope = (scope as any).block ? scope : null;
      const instanceScope = new Scope(parentScope, statement);
      instanceScope.$instance = this;

      instructions.push(
        new Instruction(
          instanceScope,
          statement,
          () => {},
          () => {},
          null,
          null
        ) as unknown as NODE
      );

      instructions.push(
        new Instruction(
          instanceScope,
          statement,
          () => {},
          () => {},
          null,
          null
        ) as unknown as NODE
      );
    }

    return instructions.length > 0 ? instructions : null;
  }

  graph(): NODE[] | null {
    this.class.declarations[this.id] = this;
    return null;
  }
}

export default BLOCK$CLASS;

File: cpp/src/nuc/BLOCK$CLASS.cpp
#include "nucleoid/nuc/BLOCK$CLASS.hpp"
#include "nucleoid/nuc/BLOCK$INSTANCE.hpp"
#include "nucleoid/Scope.hpp"
#include "nucleoid/Instruction.hpp"
#include <cstdint>
#include <random>
#include <sstream>

namespace nucleoid {

static std::string makeUUID() {
    static std::mt19937_64 rng(std::random_device{}());
    std::uniform_int_distribution<uint64_t> dist;
    std::ostringstream ss;
    ss << std::hex << dist(rng) << dist(rng);
    return ss.str();
}

NODE* BLOCK$CLASS::run(Scope* scope) {
    if (!classRef) return nullptr;

    std::vector<NODE*> toRun;

    // Determine which instances to iterate.
    std::vector<NODE*> instances;
    if (scope->instance) {
        instances.push_back(scope->instance);
    } else {
        for (auto& [k, v] : classRef->instances) {
            if (v) instances.push_back(v);
        }
    }

    for (NODE* inst : instances) {
        auto* blk = new BLOCK$INSTANCE(makeUUID());
        blk->instance    = inst;
        blk->declaration = this;
        blk->body        = statements;

        Scope* instanceScope = new Scope(scope);
        instanceScope->instance = inst;
        instanceScope->block    = blk;

        // We push the BLOCK$INSTANCE directly; Stack will expand it.
        toRun.push_back(blk);
    }

    // Return first; stack expansion handles the rest.
    return toRun.empty() ? nullptr : toRun.front();
}

} // namespace nucleoid
